src/test/java/org/cryptomator/fusecloudaccess/CloudAccessFSTest.java

```diff
@@ -44,26 +61,336 @@ public void testInterrupt() throws InterruptedException {
+		}
+
+		@DisplayName("readdir() returns EIO on any other exception")
+		@Test
+		public void testAnyExceptionReturnsEIO() {
+			FuseFillDirImpl filler = FuseFillDirImpl.getENOMEMFiller();
+
+			Mockito.when(provider.listExhaustively(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new Exception()));
+
+			Assertions.assertEquals(-ErrorCodes.EIO(), cloudFs.readdir(PATH.toString(), buf, filler, OFFSET, fi));
+		}
+	}
+
+	static class OpenTest {
+
+		private CloudProvider provider;
+		private OpenFileFactory fileFactory;
+		private CloudAccessFS cloudFs;
+		private TestFileInfo fi;
+
+		@BeforeEach
+		public void setup() {
+			provider = Mockito.mock(CloudProvider.class);
+			fileFactory = Mockito.mock(OpenFileFactory.class);
+			cloudFs = new CloudAccessFS(provider, CloudAccessFSTest.TIMEOUT, fileFactory);
+			fi = TestFileInfo.create();
+		}
+
+		@DisplayName("open() returns 0 in success and writes the handle to field FileInfo.fh")
+		@Test
+		public void testSuccessfulOpenReturnsZeroAndStoresHandle() {
+			long expectedHandle = 1337;
+
+			Mockito.when(fileFactory.open(Mockito.any(Path.class), Mockito.anySet())).thenReturn(1337L);
+
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.completedFuture(CloudItemMetadataProvider.ofPath(PATH)));
+			Assertions.assertEquals(0, cloudFs.open(PATH.toString(), fi));
+			Assertions.assertEquals(expectedHandle, fi.fh.get());
+		}
+
+		@DisplayName("open() returns ECANCELED on ClosedOpenFileFactoryException.")
+		@Test
+		public void testClosedFileFactoryExceptionReturnsECANCELED() {
+			Mockito.when(fileFactory.open(Mockito.any(Path.class), Mockito.anySet())).thenThrow(ClosedOpenFileFactoryException.class);
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.completedFuture(CloudItemMetadataProvider.ofPath(PATH)));
+			Assertions.assertEquals(-ErrorCodes.ECANCELED(), cloudFs.open(PATH.toString(), fi));
+
+		}
+
+
+		@DisplayName("open() returns ENOENT if the specified path is not found")
+		@Test
+		public void testNotFoundExceptionReturnsENOENT() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new NotFoundException()));
+			Assertions.assertEquals(-ErrorCodes.ENOENT(), cloudFs.open(PATH.toString(), fi));
+		}
+
+		@DisplayName("open() returns EIO if an IO error occured at the provider side")
+		@Test
+		public void testCloudProviderExceptionReturnsEIO() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new CloudProviderException()));
+			Assertions.assertEquals(-ErrorCodes.EIO(), cloudFs.open(PATH.toString(), fi));
+		}
+
+		@DisplayName("open() returns generic EIO if any exception is thrown")
+		@Test
+		public void testAnyExceptionReturnsEIO() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new Exception()));
+			Assertions.assertEquals(-ErrorCodes.EIO(), cloudFs.open(PATH.toString(), fi));
+		}
```

Comment:
collapse to single ParameterizedTest?

Reply:
resolved by da080597d67cce68cd7f985d6edbd924bb785bee